src/java/org/apache/cassandra/cql3/Validation.java

```diff
@@ -64,4 +70,28 @@ public static void validateKey(TableMetadata metadata, ByteBuffer key)
             throw new InvalidRequestException(e.getMessage());
         }
     }
+
+    public static void validateKeyAndCheckConstraints(TableMetadata metadata, ByteBuffer key)
+    {
+        validateKey(metadata, key);
+
+        List<ColumnMetadata> partitionKeys = metadata.partitionKeyColumns();
+        List<ColumnConstraint> partitionKeyConstraints = new ArrayList<>(partitionKeys.size());
```

Comment:
Could not be partitionKeyConstraints already "cached" in `metadata`? Because, if I am getting that right, every single time we go to validate this, basically every single time we deal with ModificationStatement.addUpdates, we are creating new array list etc ...

Reply:
Good catch. Moving it in there.